addons/purchase_stock/models/res_partner.py

```diff
@@ -20,12 +21,20 @@ def _compute_on_time_rate(self):
             ('partner_id', 'in', self.ids),
             ('date_order', '>', fields.Date.today() - timedelta(365)),
             ('qty_received', '!=', 0),
-        ]).filtered(lambda l: l.product_id.sudo().product_tmpl_id.type != 'service' and l.order_id.state in ['done', 'purchase'])
+            ('product_id.product_tmpl_id.type', '!=', 'service'),
+            ('order_id.state', 'in', ['done', 'purchase'])
+        ])
+        lines_qty_done = defaultdict(lambda: 0)
+        moves = self.env['stock.move'].search([
+            ('id', 'in', order_lines.move_ids.ids),
+            ('state', '=', 'done')]).filtered(lambda m: m.date.date() <= m.purchase_line_id.date_planned.date())
```

Comment:
Maybe add a filter in the search to reduce the record fetch from database (and reduce the python filtered 'load')
`('date', '<=', m.purchase_line_id.date_planned + timedelta(day=1))`
(adding the `+ timedelta(day=1)` because we compare date and then avoid to filtered out to much)